### ISSUE TITLE
container statistics  (docker 1.11.1 / ecs agent 1.10.0)

### ISSUE THREAD

**Issue by leonblueconic** [aws/amazon-ecs-agent#416]
We are running 3 servers on the the 2016.03.c-amazon-ecs-optimized AMI and only one off them fails to report statistics. In the logs I can see the following messages, which look related:

> 2016-06-03T09:36:53Z [DEBUG] Connecting to TCS endpoint https://ecs-t-2.eu-west-1.amazonaws.com/ module="tcs handler"
2016-06-03T09:36:53Z [INFO] Creating poll dialer module="ws client" host="ecs-t-2.eu-west-1.amazonaws.com"
2016-06-03T09:36:53Z [DEBUG] Starting websocket poll loop module="tcs client"
2016-06-03T09:37:10Z [DEBUG] Got a message from websocket module="ws client" message="{"type":"HeartbeatMessage","message":{"healthy":true}}"
2016-06-03T09:37:10Z [DEBUG] ACS activity occured
2016-06-03T09:37:53Z [ERROR] Error getting message from ws backend module="ws client" err="websocket: close 1002 Channel long idle: No message is received, close the channel"
2016-06-03T09:37:53Z [INFO] Error from tcs; backing off module="tcs handler" err="websocket: close 1002 Channel long idle: No message is received, close the channel"
2016-06-03T09:38:10Z [DEBUG] Got a message from websocket module="ws client" message="{"type":"HeartbeatMessage","message":{"healthy":true}}"
2016-06-03T09:38:10Z [DEBUG] ACS activity occured
2016-06-03T09:39:02Z [DEBUG] Connecting to TCS endpoint https://ecs-t-2.eu-west-1.amazonaws.com/ module="tcs handler"
2016-06-03T09:39:02Z [INFO] Creating poll dialer module="ws client" host="ecs-t-2.eu-west-1.amazonaws.com"
2016-06-03T09:39:02Z [DEBUG] Starting websocket poll loop module="tcs client"

and

> 2016-06-03T09:18:13Z [DEBUG] Got a message from websocket module="ws client" message="{"message":{},"type":"AckPublishMetric"}"
2016-06-03T09:18:13Z [DEBUG] Received AckPublishMetric from tcs module="tcs handler"
2016-06-03T09:18:13Z [DEBUG] Got a message from websocket module="ws client" message="{"message":{},"type":"AckPublishMetric"}"
2016-06-03T09:18:13Z [DEBUG] Received AckPublishMetric from tcs module="tcs handler"
2016-06-03T09:18:13Z [ERROR] Error getting message from ws backend module="ws client" err="websocket: close 1008 InvalidContent : Unexpected text frame received"
2016-06-03T09:18:13Z [INFO] Error from tcs; backing off module="tcs handler" err="websocket: close 1008 InvalidContent : Unexpected text frame received"
2016-06-03T09:18:18Z [DEBUG] Connecting to TCS endpoint https://ecs-t-2.eu-west-1.amazonaws.com/ module="tcs handler"

**Comment by samuelkarp** [comment]
The presence of `AckPublishMetric` typically indicates that metrics are getting pushed to ECS successfully.  Can you explain a bit more about what specifically you're observing?  Are you lacking metrics on the cluster level or service level?

**Comment by leonblueconic** [comment]
Metrics are lacking on the service level. Below are 2 graphs for two service one is running on the 'faulty' ecs instance the other one reports fine. Both are running on the same AMI, the same docker version and the same ecs agent. 
![cpu-capture](https://cloud.githubusercontent.com/assets/19727725/15870095/64aa7752-2cef-11e6-99b6-93b9423f1b5a.PNG)
![cpu-capture-working](https://cloud.githubusercontent.com/assets/19727725/15870096/64b53fe8-2cef-11e6-9fec-e64a5df21f9b.PNG)

Metrics report for the cpuutilization appear to be correct

**Comment by samuelkarp** [comment]
Thanks for the details.  I'm wondering if something might be going odd with metrics processing in ECS for your service; can you provide the ARN of the service and cluster that is affected?  If you are not comfortable posting this information publicly, you can email it directly to me at skarp (at) amazon.com.

**Comment by leonblueconic** [comment]
I have provided some details thru email.

**Comment by samuelkarp** [comment]
Thank you.  We'll take a look at this and see what we can find.

**Comment by samuelkarp** [comment]
Can you provide some sample payloads that you see in the logs for the affected instance?  You'll see them at debug level in the logs, looking something like this:

```
2016-06-07T23:33:39Z [DEBUG] sending payload module="tcs client" payload="{"type":"PublishMetricsRequest","message":{"metadata":{"cluster":"default","containerInstance":"arn:aws:ecs:us-west-2:123456789012:container-instance/48bba1e6-d37e-4b0c-b73a-a2a28dfed726","fin":true,"idle":false,"messageId":"092740f4-c3f0-4a79-8eed-18f61a842807"},"taskMetrics":[{"containerMetrics":[{"cpuStatsSet":{"max":0.07654689252376556,"min":0.06465789675712585,"sampleCount":19,"sum":1.3596976920962334},"memoryStatsSet":{"max":4,"min":4,"sampleCount":20,"sum":80}}],"taskArn":"arn:aws:ecs:us-west-2:123456789012:task/bc2185ad-77b5-4b18-ac8a-ca6c21e890f0","taskDefinitionFamily":"my_task_def","taskDefinitionVersion":"1"}],"timestamp":1465342419}}"
```

As before, if you're not comfortable posting this publicly you can email it directly to me.

**Comment by leonblueconic** [comment]
Sent a followup email due to the sensitive nature of information contained in the logs.

**Comment by samuelkarp** [comment]
@leonblueconic Thanks for sending the logs.  From what we can tell so far, there are two things that look weird:
* An invalid payload was received at `2016-06-03T09:18:13.535Z`.
* During most of the other time, the instance does not appear to be sending payloads.  The payloads we are receiving are showing up in the metrics you can see.

We're not sure of a root cause here, but to start ruling things out, I'd like to ask for more information about your specific configuration.  Again, if you're not comfortable answering these questions publicly, you can email the answers directly to me.  Specifically, can you provide:
* Confirmation that the configuration on this instance is the same as the configuration on your other instances that appear to be working
* Confirmation of whether or not `docker stats` for that container provides data (we rely on `docker stats`, so if it's not providing data or only working intermittently that would definitely be a big clue)
* Whether you've already tried restarting the ECS agent, restarting Docker, or moving the service to a different host in the same cluster and whether that helps

**Comment by leonblueconic** [comment]
* We launch our ecs instances in an autoscaling group. Of the 3 servers launched in this group (at the the same time) only one shows this problems.

* `docker stats` seems to be working fine. 

* ECS agent was restarted serveral times to correct this problem. We haven't tried to restart docker on this host.  When we migrate a service to another instance the cloudwatch metrics start to work and we don't see any gaps in the reporting.

**Comment by richardpen** [comment]
@leonbluconic Thanks for providing this information. We were unable to root-cause the issue from the log files and the information that you shared with us so far. Could you please provide us the following additional information to help us root-cause this issue?
 
* Run `docker kill -s USR1 ecs-agent` while you have the agent set with `ECS_LOGLEVEL=debug` and send us the resulting stack trace (which will be output to the latest agent log file under `/var/log/ecs`).
* Restart the agent again, let it run for an hour or so and send the logs from that time period including when you restarted the agent (I'm looking for useful output from when the agent starts up).
* Send the Docker daemon logs located under `/var/log/docker`.

If you're not comfortable posting this publicly, you can email me at penyin (at) amazon.com.
 
Thanks.

**Comment by leonblueconic** [comment]
Over the weekend I did provide some logs through private email.

**Comment by samuelkarp** [comment]
@leonblueconic Thanks for sharing all this information.  There's definitely some sort of bug here; I can see another invalid payload generated at `2016-06-12T21:02:48Z`.  We'll keep investigating to see if we can figure out exactly what's going on.

**Comment by UnquietCode** [comment]
I am experiencing a similar issue, where we have multiple servers in an ASG, but some metrics are not being reported from one server. Since we use the [Data Samples trick](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/cloudwatch-metrics.html#cw_running_task_count) to keep tabs on how many of our service tasks are running, we end up underreporting how many active tasks there are.

**Comment by aaithal** [comment]
@leonblueconic I wanted to provide you an update of where we are with our investigation of this issue: 
 
1. We have found and fixed a [potential bug in the agent](https://github.com/aws/amazon-ecs-agent/pull/430), which would explain the invalid payload messages that were present in the logs that you shared with us 
2. We found another bug, where errors with sending metric message payloads to ECS were being swallowed in the agent and have [fixed that as well](https://github.com/aws/amazon-ecs-agent/pull/425)
3. The logs also indicate that the `docker stats` API that we use to get container metrics has been hanging for well over 11457 minutes. We think that periodically closing and reopening the `stats` channel should solve this issue
 
The first two bug fixes are available in the [latest version of the ECS Agent](https://github.com/aws/amazon-ecs-agent/releases/tag/v1.11.0) released on 07/14/2016. We will let you know when we have more details about the last one.
 
Thanks,
Anirudh

**Comment by aaithal** [comment]
@UnquietCode as per my [previous post](https://github.com/aws/amazon-ecs-agent/issues/416#issuecomment-232717718) to @leonblueconic, we have identified a number of potential bugs that could be leading up to this issue. It would really help us if you could provide more details regarding the issue that you're facing:

1. We would like the agent's stack trace from an instance where you suspect you're seeing this issue. You'd be able to collect this if the ECS Agent was started with `ECS_LOGLEVEL=debug` by running the command `docker kill -s USR1 ecs-agent`
2. The cluster and service arn where you're seeing this issue
3. Agent logs (from `/var/log/ecs`)

You can send these to aithal at amazon dot com if you do not want to post these in public.

Thanks,
Anirudh

**Comment by leonblueconic** [comment]
I will try the new version a soon as the rpm for this version is pushed to the Amazon Linux package repository.

**Comment by leonblueconic** [comment]
ECS agents where updated. On one of our servers we get the following errors:

```
2016-07-15T07:13:27Z [DEBUG] Got a message from websocket
2016-07-15T07:13:27Z [DEBUG] Received message of type: HeartbeatMessage
2016-07-15T07:13:27Z [DEBUG] ACS activity occured
2016-07-15T07:14:03Z [DEBUG] Connecting to TCS endpoint https://ecs-t-2.eu-west-1.amazonaws.com/ module="tcs handler"
2016-07-15T07:14:03Z [INFO] Creating poll dialer, host: ecs-t-2.eu-west-1.amazonaws.com
2016-07-15T07:14:03Z [DEBUG] Starting websocket poll loop module="tcs client"
2016-07-15T07:14:03Z [WARN] Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"
2016-07-15T07:14:03Z [WARN] Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"
2016-07-15T07:14:03Z [WARN] Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"
2016-07-15T07:14:03Z [WARN] Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"
2016-07-15T07:14:23Z [WARN] Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"
2016-07-15T07:14:23Z [WARN] Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"
2016-07-15T07:14:23Z [WARN] Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"
2016-07-15T07:14:23Z [WARN] Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"
2016-07-15T07:14:27Z [DEBUG] Got a message from websocket
2016-07-15T07:14:27Z [DEBUG] Received message of type: HeartbeatMessage
2016-07-15T07:14:27Z [DEBUG] ACS activity occured
2016-07-15T07:14:43Z [WARN] Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"
2016-07-15T07:14:43Z [WARN] Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"
2016-07-15T07:14:43Z [WARN] Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"
2016-07-15T07:14:43Z [WARN] Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"
2016-07-15T07:15:03Z [ERROR] Error getting message from ws backend: websocket: close 1002 Channel long idle: No message is received, close the channel
2016-07-15T07:15:03Z [WARN] Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"
2016-07-15T07:15:03Z [WARN] Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"
2016-07-15T07:15:03Z [WARN] Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"
2016-07-15T07:15:03Z [WARN] Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"
2016-07-15T07:15:03Z [INFO] Error from tcs; backing off module="tcs handler" err="websocket: close 1002 Channel long idle: No message is received, close the channel"


```

**Comment by aaithal** [comment]
@leonblueconic thanks for providing these logs. The modifications in the latest version of ECS Agent seem to be working in that the error message is being logged now (which wasn't the case in the previous ECS Agent Version). This also helps us in ruling out any issue with our metrics backed as this seems to be related to some sort of a serialization issue in the ECS Agent. We are trying to reproduce this in our test setup. We will let you know when we have an update this.

**Comment by aaithal** [comment]
After updating the ECS Agent on my instance and running a high-CPU workload, I have started seeing more of these messages:
```
./ecs-agent.log.2016-07-25-19:2016-07-25T19:59:46Z [DEBUG] Time since last stat is zero. Ignoring cpu stat. lastStat: {4.3041987 44 2016-07-25 19:59:46 +0000 UTC 20442827875} rawStat: &{20448869533 46346240 {63605073586 0 0xdd8ae0}} stat: {NaN 44 2016-07-25 19:59:46 +0000 UTC 20448869533} 
./ecs-agent.log.2016-07-25-19:2016-07-25T19:59:46Z [DEBUG] Time since last stat is zero. Ignoring cpu stat. lastStat: {0.12619492 24 2016-07-25 19:59:46 +0000 UTC 1584195850} rawStat: &{1584648929 25997312 {63605073586 0 0xdd8ae0}} stat: {NaN 24 2016-07-25 19:59:46 +0000 UTC 1584648929} 
./ecs-agent.log.2016-07-25-19:2016-07-25T19:59:46Z [DEBUG] Time since last stat is zero. Ignoring cpu stat. lastStat: {0.122083664 24 2016-07-25 19:59:46 +0000 UTC 1041979414} rawStat: &{1042403364 26099712 {63605073586 0 0xdd8ae0}} stat: {NaN 24 2016-07-25 19:59:46 +0000 UTC 1042403364} 
./ecs-agent.log.2016-07-25-19:2016-07-25T19:59:46Z [DEBUG] Time since last stat is zero. Ignoring cpu stat. lastStat: {0.13177656 24 2016-07-25 19:59:46 +0000 UTC 1333420387} rawStat: &{1333815848 25980928 {63605073586 0 0xdd8ae0}} stat: {NaN 24 2016-07-25 19:59:46 +0000 UTC 1333815848} 
./ecs-agent.log.2016-07-25-19:2016-07-25T19:59:46Z [DEBUG] Time since last stat is zero. Ignoring cpu stat. lastStat: {3.5063858 44 2016-07-25 19:59:46 +0000 UTC 23166187127} rawStat: &{23173051115 46342144 {63605073586 0 0xdd8ae0}} stat: {NaN 44 2016-07-25 19:59:46 +0000 UTC 23173051115} 
```
Seems that`docker stats` is returning the same timestamp across multiple stats messages. Looking more into it, `docker stats` seems to be returning metrics only by `second` granularity, not mills/nanos as per [documentation](https://docs.docker.com/engine/reference/api/docker_remote_api_v1.20/#get-container-stats-based-on-resource-usage), which seems to the root cause of this issue. 

Digging more into docker/docker code-base, it seems to be a regression that was introduced with this [commit](https://github.com/docker/docker/commit/9c4570a958df42d1ad19364b1a8da55b891d850a#diff-b56258ecab11f7b797306907327cc904R1086). A recent commit seems to have overwritten this with [protobuf timestamp](https://github.com/docker/docker/commit/29b2714580d085533c29807fa337c2b7a302abb6). We are not yet sure if that fixes the issue though. We'll give that a try and update the thread.

**Comment by aaithal** [comment]
Definitely a regression/change of behavior in docker.

I wrote a sample program to invoke `docker stats` and emit the payload received from docker. With docker v1.9.1:

```
stat:  &{2016-07-25 22:09:07.305488472 +0000 UTC {0} {0 0 0 0 0 0 0 0} map[eth0:{0 648 0 8 0 8 0 648}] {{0 6365184 4362240 5898240 1480 1384448 4362240 0 0 3372 0 46 1384448 0 0 12288 0 9223372036854771712 2340 421888 1417216 1417216 1480 6365184 12288 421888 2340 5898240 3372 0 0} 7815168 7749632 0 4142559232} {[{202 0 Read 6414336} {202 0 Write 0} {202 0 Sync 0} {202 0 Async 6414336} {202 0 Total 6414336}] [{202 0 Read 210} {202 0 Write 0} {202 0 Sync 0} {202 0 Async 210} {202 0 Total 210}] [] [] [] [] [] []} {{[21664308 30357254] 30000000 52021562 0} 29406001770000000 {0 0 0}} {{[] 0 0 0} 0 {0 0 0}}}
stat:  &{2016-07-25 22:09:08.305523477 +0000 UTC {0} {0 0 0 0 0 0 0 0} map[eth0:{0 648 0 8 0 8 0 648}] {{0 6365184 4362240 5898240 1480 1384448 4362240 0 0 3372 0 46 1384448 0 0 12288 0 9223372036854771712 2340 421888 1417216 1417216 1480 6365184 12288 421888 2340 5898240 3372 0 0} 7815168 7749632 0 4142559232} {[{202 0 Read 6414336} {202 0 Write 0} {202 0 Sync 0} {202 0 Async 6414336} {202 0 Total 6414336}] [{202 0 Read 210} {202 0 Write 0} {202 0 Sync 0} {202 0 Async 210} {202 0 Total 210}] [] [] [] [] [] []} {{[21664308 30357254] 30000000 52021562 0} 29406003760000000 {0 0 0}} {{[21664308 30357254] 30000000 52021562 0} 29406001770000000 {0 0 0}}}
stat:  &{2016-07-25 22:09:09.305449475 +0000 UTC {0} {0 0 0 0 0 0 0 0} map[eth0:{0 648 0 8 0 8 0 648}] {{0 6365184 4362240 5898240 1480 1384448 4362240 0 0 3372 0 46 1384448 0 0 12288 0 9223372036854771712 2340 421888 1417216 1417216 1480 6365184 12288 421888 2340 5898240 3372 0 0} 7815168 7749632 0 4142559232} {[{202 0 Read 6414336} {202 0 Write 0} {202 0 Sync 0} {202 0 Async 6414336} {202 0 Total 6414336}] [{202 0 Read 210} {202 0 Write 0} {202 0 Sync 0} {202 0 Async 210} {202 0 Total 210}] [] [] [] [] [] []} {{[21664308 30357254] 30000000 52021562 0} 29406005770000000 {0 0 0}} {{[21664308 30357254] 30000000 52021562 0} 29406003760000000 {0 0 0}}}
stat:  &{2016-07-25 22:09:10.305440481 +0000 UTC {0} {0 0 0 0 0 0 0 0} map[eth0:{0 648 0 8 0 8 0 648}] {{0 6365184 4362240 5898240 1480 1384448 4362240 0 0 3372 0 46 1384448 0 0 12288 0 9223372036854771712 2340 421888 1417216 1417216 1480 6365184 12288 421888 2340 5898240 3372 0 0} 7815168 7749632 0 4142559232} {[{202 0 Read 6414336} {202 0 Write 0} {202 0 Sync 0} {202 0 Async 6414336} {202 0 Total 6414336}] [{202 0 Read 210} {202 0 Write 0} {202 0 Sync 0} {202 0 Async 210} {202 0 Total 210}] [] [] [] [] [] []} {{[21664308 30357254] 30000000 52021562 0} 29406007770000000 {0 0 0}} {{[21664308 30357254] 30000000 52021562 0} 29406005770000000 {0 0 0}}}
stat:  &{2016-07-25 22:09:11.305489343 +0000 UTC {0} {0 0 0 0 0 0 0 0} map[eth0:{0 648 0 8 0 8 0 648}] {{0 6365184 4362240 5898240 1480 1384448 4362240 0 0 3372 0 46 1384448 0 0 12288 0 9223372036854771712 2340 421888 1417216 1417216 1480 6365184 12288 421888 2340 5898240 3372 0 0} 7815168 7749632 0 4142559232} {[{202 0 Read 6414336} {202 0 Write 0} {202 0 Sync 0} {202 0 Async 6414336} {202 0 Total 6414336}] [{202 0 Read 210} {202 0 Write 0} {202 0 Sync 0} {202 0 Async 210} {202 0 Total 210}] [] [] [] [] [] []} {{[21664308 30357254] 30000000 52021562 0} 29406009770000000 {0 0 0}} {{[21664308 30357254] 30000000 52021562 0} 29406007770000000 {0 0 0}}}
```

With docker v1.11.1 (Notice how we loose the millisecond precision here):
```
stat:  &{2016-07-25 22:07:23 +0000 UTC {0} {0 0 0 0 0 0 0 0} map[eth0:{0 648 0 8 0 8 0 648}] {{0 40960 4096 8192 293 1409024 4096 0 0 647 0 0 1409024 0 0 4096 0 18446744073709551615 2189 0 1437696 1437696 293 40960 4096 0 2189 8192 647 0 0} 3538944 1687552 0 16827506688} {[] [] [] [] [] [] [] []} {{[6218023 5911472 820045 945660] 0 13895200 0} 11085180000000 {0 0 0}} {{[] 0 0 0} 0 {0 0 0}}}
stat:  &{2016-07-25 22:07:24 +0000 UTC {0} {0 0 0 0 0 0 0 0} map[eth0:{0 648 0 8 0 8 0 648}] {{0 40960 4096 8192 293 1409024 4096 0 0 647 0 0 1409024 0 0 4096 0 18446744073709551615 2189 0 1437696 1437696 293 40960 4096 0 2189 8192 647 0 0} 3538944 1687552 0 16827506688} {[] [] [] [] [] [] [] []} {{[6218023 5911472 820045 945660] 0 13895200 0} 11089180000000 {0 0 0}} {{[6218023 5911472 820045 945660] 0 13895200 0} 11085180000000 {0 0 0}}}
stat:  &{2016-07-25 22:07:25 +0000 UTC {0} {0 0 0 0 0 0 0 0} map[eth0:{0 648 0 8 0 8 0 648}] {{0 40960 4096 8192 293 1409024 4096 0 0 647 0 0 1409024 0 0 4096 0 18446744073709551615 2189 0 1437696 1437696 293 40960 4096 0 2189 8192 647 0 0} 3538944 1687552 0 16827506688} {[] [] [] [] [] [] [] []} {{[6218023 5911472 820045 945660] 0 13895200 0} 11093170000000 {0 0 0}} {{[6218023 5911472 820045 945660] 0 13895200 0} 11089180000000 {0 0 0}}}
stat:  &{2016-07-25 22:07:26 +0000 UTC {0} {0 0 0 0 0 0 0 0} map[eth0:{0 648 0 8 0 8 0 648}] {{0 40960 4096 8192 293 1409024 4096 0 0 647 0 0 1409024 0 0 4096 0 18446744073709551615 2189 0 1437696 1437696 293 40960 4096 0 2189 8192 647 0 0} 3538944 1687552 0 16827506688} {[] [] [] [] [] [] [] []} {{[6218023 5911472 820045 945660] 0 13895200 0} 11097170000000 {0 0 0}} {{[6218023 5911472 820045 945660] 0 13895200 0} 11093170000000 {0 0 0}}}
```

* This issue exists even with the latest docker build from master (using [this commit](https://github.com/docker/docker/commit/8f11896a45aea554935353cb1347cc2aa2529395))
* Another issue here is the increased frequency of stats being reported by `docker stats` under high load.

Here's the sample program that I wrote for this:
```go
  package main

  import (
          "fmt"

          docker "github.com/fsouza/go-dockerclient"
  )

  func main() {
          endpoint := "unix:///var/run/docker.sock"
          client, _ := docker.NewClient(endpoint)
          stats := make(chan *docker.Stats)
          cancel := make(chan bool)
          id := "074985022d7c"
          options := docker.StatsOptions{
                  ID:     id,
                  Stats:  stats,
                  Stream: true,
                  Done:   cancel,
          }

          statsComplete := make(chan struct{})
          go func() {
                  statsErr := client.Stats(options)
                  if statsErr != nil {
                          fmt.Printf("Error retrieving stats for container %s: %v\n", id, statsErr)
                  }
                  close(statsComplete)
          }()

          go func() {
                  select {
                  case <-statsComplete:
                  }
          }()

          for s := range stats {
                  fmt.Println("stat: ", s)
          }
  }
```

**Comment by aaithal** [comment]
We've opened a [new issue](https://github.com/docker/docker/issues/25078) in docker for this. Meanwhile, the fix that we made in the agent as per #459 should ensure that metrics with duplicate timestamps are not processed by the ECS Agent.

**Comment by leonblueconic** [comment]
Thanks for the support so far. Looking forward to the new release.

**Comment by steevel** [comment]
My logs are also full with what @leonblueconic reported:
Error publishing metrics module="tcs client" err="Could not marshal Request (PublishMetricsRequest): json: error calling MarshalJSON for type json.RawMessage: invalid character '+' looking for beginning of value"

Ping me if I can help with debug somehow or provide more logs.

**Comment by aaithal** [comment]
@steevel Thanks for offering to help us with the investigation. As per my earlier response, we root-caused this to a regression in docker v1.11.2 and have reported the [same upstream](https://github.com/docker/docker/issues/25078). We have also pushed a fix for this (#459) in the `dev` branch of the agent, which will make sure that metrics with duplicate timestamps are ignored by the ECS Agent.

**Comment by henryturk** [comment]
Anyone know of a reliable way of downgrading docker to 1.11.1 using launch configurations/user data on the latest ECS AMI? This issue is wreaking havoc on my auto scaling policies.

**Comment by samuelkarp** [comment]
This should be fixed with our [v1.11.1](https://github.com/aws/amazon-ecs-agent/releases/tag/v1.11.1) release.
